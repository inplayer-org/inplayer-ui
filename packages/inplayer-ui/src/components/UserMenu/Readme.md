The UserMenu is a component which renders an overlay dropdown and is used for rendering a profile image, user name and user-related links.

```jsx static
import { UserMenu } from 'inplayer-ui';

function Page(props) {
  return (
    <UserMenu
      image="https://www.impaact4tb.org/wp-content/uploads/2018/07/Generic-male-icon.png"
      menuItems={[{ title: 'Hello' }]}
    />
  );
}
```

### Standard menu

```jsx
<UserMenu
  image="https://www.impaact4tb.org/wp-content/uploads/2018/07/Generic-male-icon.png"
  menuTitle="User name"
  menuItems={[{ title: 'Hello' }]}
/>
```

### With special action item at bottom

```jsx
<UserMenu
  image="https://www.impaact4tb.org/wp-content/uploads/2018/07/Generic-male-icon.png"
  menuTitle='User name'
  menuItems={[{ title: 'Hello' }, { title: 'World' }]}
  actionItem={{ title: 'Logout' }}
/>
```
