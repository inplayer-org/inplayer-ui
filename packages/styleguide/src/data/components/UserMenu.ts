import { ComponentData } from "../types";

const usage = `
  import { UserMenu } from 'inplayer-ui';

  function Page(props) {
    return (
      <UserMenu
        image="https://www.impaact4tb.org/wp-content/uploads/2018/07/Generic-male-icon.png"
        menuItems={[{ title: 'Hello' }]}
      />
    );
  }
`;

const standardMenu = `<UserMenu
  image="https://www.impaact4tb.org/wp-content/uploads/2018/07/Generic-male-icon.png"
  menuTitle="User name"
  menuItems={[{ title: 'Hello' }]}
/>
`;

const specialActionItem = `<UserMenu
  image="https://www.impaact4tb.org/wp-content/uploads/2018/07/Generic-male-icon.png"
  menuTitle='User name'
  menuItems={[{ title: 'Hello' }, { title: 'World' }]}
  actionItem={{ title: 'Logout' }}
/>
`;

const UserMenu: ComponentData = {
  description: `
    The UserMenu is a component which renders an overlay dropdown and is used for rendering a profile image, user name and user-related links.
  `,
  path: 'src/components/UserMenu/index.js',
  usage,
  examples: [
    {
      title: 'Standard Menu',
      code: standardMenu,
    },
    {
      title: 'With special action item at bottom',
      code: specialActionItem,
    },
  ],
  propsAndMethods: [
    { name: 'image', type: 'union', default: 'Required' },
    { name: 'className', type: 'string' },
    { name: 'menuTitle', type: 'string' },
    { name: 'style', type: 'Object', default: '{ }'},
  ],
}

export default UserMenu;