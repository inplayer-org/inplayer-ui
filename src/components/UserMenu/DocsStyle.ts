import styled from 'styled-components';

export const UserMenuWrapper = styled.div`
  width: 90%;
  height: 240px;
`;
