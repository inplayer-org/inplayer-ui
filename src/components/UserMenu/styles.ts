import styled from 'styled-components';
import { ifProp } from 'styled-tools';
import colors from '../../theme/colors';

export const ProfileImage = styled.img`
  border-radius: 50%;
  height: 30px;
  width: 30px;
`;

export const MenuButton = styled.div`
  align-items: center;
  background: transparent;
  border-radius: 50%;
  display: flex;
  height: 30px;
  justify-content: center;
  min-width: 30px;
  transition: all 0.5s ease;

  &:hover {
    background: ${colors.lightGray};
  }
`;

export const MenuArrow = styled.i<{ open: boolean }>`
  border: solid ${colors.darkGray};
  border-width: 1px 0 0 1px;
  display: inline-block;
  padding: 4px;
  transform: ${ifProp('open', 'rotate(45deg)', 'rotate(-135deg)')};
  transition: all 0.5s ease;
`;

export const UserMenuContainer = styled.div`
  align-items: center;
  display: flex;
  justify-content: flex-end;
  max-width: 100%;
  position: relative;
  padding: 0 10px;
`;

export const CloseLayer = styled.div`
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1;
`;

export const MenuTitleContainer = styled.span`
  cursor: pointer;
  display: flex;
  align-items: baseline;
`;

export const MenuTitle = styled.span`
  color: ${({ theme }) => theme.palette.text.light};
  transition: color 0.3s ease;
  font-size: ${({ theme }) => theme.font.sizes.medium};
  padding: 0 3px 0 10px;
  cursor: pointer;

  &:hover {
    color: ${({ theme }) => theme.palette.text.main};
  }
`;
