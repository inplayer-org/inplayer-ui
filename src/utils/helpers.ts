import moment from 'moment';

export const getMonthOptions = (): Array<any> =>
  moment.months().map((label, value) => ({ value, displayName: label }));

export const getYearOptions = () => {
  const yearOptions: Array<any> = [];
  for (let i = moment().year() - 100; i <= moment().year(); i += 1) {
    yearOptions.push({ value: i.toString(), displayName: i.toString() });
  }

  return yearOptions;
};
