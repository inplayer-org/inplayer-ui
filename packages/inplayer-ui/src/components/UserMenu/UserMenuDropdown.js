// @flow
import React from 'react';
import styled, { css } from 'styled-components';
import { ifProp } from 'styled-tools';
import { transparentize } from 'polished';
import { uiColors, fontWeights, fontSizes } from 'utils';
import colors from 'config/colors';

const Container = styled.div`
  align-items: flex-start;
  box-shadow: 0 1px 5px 0 ${transparentize(0.8, '#000')};
  z-index: 1000;
  min-width: 100%;
  background: ${colors.white};
  display: inline-flex;
  flex-direction: column;
  justify-content: space-between;
  padding: 1.125rem 0;
  position: absolute;
  right: 1%;
  top: 40px;

  &::before {
    content: '';
    display: block;
    box-shadow: -1px -1px 1px 0 ${transparentize(0.9, '#000')};
    background: ${colors.white};
    transform: rotate(45deg);
    width: 10px;
    height: 10px;
    position: absolute;
    top: -5px;
    right: 11px;
  }
`;

const Item = styled.a`
  color: ${uiColors('text.light')};
  box-sizing: border-box;
  font-weight: ${fontWeights('light')};
  letter-spacing: 0.02em;
  line-height: 2;
  padding: 0.375rem 0.625rem;
  text-decoration: none;
  transition: color 0.3s ease;
  white-space: nowrap;
  width: 100%;
  font-size: ${fontSizes('medium')};

  &:hover {
    color: ${uiColors('text.main')};
  }
`;

const ActionItem = styled(Item)`
  border-top: 1px solid ${uiColors('text.disabled')};
  margin-top: 0.75rem;
  padding-top: 1.125rem;
  ${ifProp(
    'smallSize',
    css`
      font-size: ${fontSizes('small')};
      border-top: none;
    `
  )};
`;

const ItemContainer = styled.div`
  display: flex;
  align-items: center;
  width: 100%;
`;

const ItemImage = styled.img`
  height: 30px;
  width: 30px;
  border-radius: 50%;
  margin-left: 0.625rem;
`;

type MenuItem = {
  title: string,
  href?: string,
  image?: string,
  onClick?: (e: SyntheticEvent<*>) => any,
  smallSize?: boolean,
};

export type UserMenuDropdownProps = {
  menuItems: Array<MenuItem>,
  actionItem?: MenuItem,
};

const UserMenuDropdown = ({ menuItems, actionItem }: UserMenuDropdownProps) => (
  <Container>
    {menuItems.map((item, i) => (
      <ItemContainer key={i}>
        {item.image && <ItemImage src={item.image} />}
        <Item key={i} href={item.href} onClick={item.onClick}>
          {item.title}
        </Item>
      </ItemContainer>
    ))}
    {actionItem && (
      <ActionItem
        href={actionItem.href}
        onClick={actionItem.onClick}
        smallSize={actionItem.smallSize}
      >
        {actionItem.title}
      </ActionItem>
    )}
  </Container>
);

UserMenuDropdown.defaultProps = {
  actionItem: null,
};

export default UserMenuDropdown;
