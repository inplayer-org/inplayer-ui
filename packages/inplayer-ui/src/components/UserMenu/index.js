// @flow
import React, { type Element } from 'react';
import styled from 'styled-components';
import { ifProp } from 'styled-tools';
import colors from 'config/colors';

import { uiColors, fontSizes } from 'utils/index';
import UserMenuDropdown, { type UserMenuDropdownProps } from './UserMenuDropdown';

const ProfileImage = styled.img`
  border-radius: 50%;
  height: 30px;
  width: 30px;
`;

const MenuButton = styled.div`
  align-items: center;
  background: transparent;
  border-radius: 50%;
  display: flex;
  height: 30px;
  justify-content: center;
  min-width: 30px;
  transition: all 0.5s ease;

  &:hover {
    background: ${colors.lightGray};
  }
`;

const MenuArrow = styled.i`
  border: solid ${colors.darkGray};
  border-width: 1px 0 0 1px;
  display: inline-block;
  padding: 4px;
  transform: ${ifProp('open', 'rotate(45deg)', 'rotate(-135deg)')};
  transition: all 0.5s ease;
`;

const UserMenuContainer = styled.div`
  align-items: center;
  cursor: pointer;
  display: flex;
  justify-content: flex-end;
  max-width: 100%;
  position: relative;
  padding: 0 10px;
`;

const CloseLayer = styled.div`
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1;
`;

const MenuTitle = styled.span`
  color: ${uiColors('text.light')};
  transition: color 0.3s ease;
  font-size: ${fontSizes('medium')};
  padding: 0 3px 0 10px;

  &:hover {
    color: ${uiColors('text.main')};
  }
`;

export type UserMenuProps = UserMenuDropdownProps & {
  image: Element<*> | string,
  className?: string,
  style?: Object,
  menuTitle?: string,
};

type UserMenuState = {
  open: boolean,
};

class UserMenu extends React.Component<UserMenuProps, UserMenuState> {
  state = {
    open: false,
  };

  toggleMenuOpen = () => {
    this.setState(({ open }) => ({ open: !open }));
  };

  render() {
    const { image, menuItems, actionItem, className, style, menuTitle } = this.props;
    const { open } = this.state;

    return (
      <>
        <UserMenuContainer className={className} style={style} onClick={this.toggleMenuOpen}>
          {image && <ProfileImage src={image} />}
          <MenuTitle>{menuTitle}</MenuTitle>
          <MenuButton>
            <MenuArrow open={open} />
          </MenuButton>
          {open && (
            <UserMenuDropdown
              open={open}
              menuItems={menuItems}
              actionItem={actionItem}
              onClick={this.toggleMenuOpen}
            />
          )}
        </UserMenuContainer>
        {open && <CloseLayer onClick={this.toggleMenuOpen} />}
      </>
    );
  }
}

UserMenu.defaultProps = {
  className: '',
  style: {},
};

export default UserMenu;
