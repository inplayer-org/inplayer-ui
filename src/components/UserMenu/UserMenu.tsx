import React, { useState } from 'react';
import UserMenuDropdown, { Props as UserMenuDropdownProps } from './UserMenuDropdown';
import {
  CloseLayer,
  MenuArrow,
  MenuButton,
  MenuTitle,
  ProfileImage,
  UserMenuContainer,
  MenuTitleContainer,
} from './styles';

export type Props = UserMenuDropdownProps & {
  image?: string;
  menuTitle?: string;
};

const UserMenu = ({ image, menuItems, actionItem, menuTitle }: Props) => {
  const [open, setOpen] = useState(false);

  return (
    <>
      <UserMenuContainer>
        {image && <ProfileImage src={image} />}
        <MenuTitleContainer onClick={() => setOpen(!open)}>
          <MenuTitle>{menuTitle}</MenuTitle>
          <MenuButton>
            <MenuArrow open={open} />
          </MenuButton>
        </MenuTitleContainer>
        {open && (
          <UserMenuDropdown
            menuItems={menuItems}
            actionItem={actionItem}
            onClick={() => setOpen(!open)}
          />
        )}
      </UserMenuContainer>
      {open && <CloseLayer onClick={() => setOpen(!open)} />}
    </>
  );
};
export default UserMenu;
