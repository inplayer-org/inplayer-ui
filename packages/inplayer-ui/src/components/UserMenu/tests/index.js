import { generateSnapshots } from 'packages/inplayer-ui/src/components/UserMenu/tests/helpers';

import UserMenu from '../index';

generateSnapshots(UserMenu);
