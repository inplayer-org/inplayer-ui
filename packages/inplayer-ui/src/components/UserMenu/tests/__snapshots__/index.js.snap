// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserMenu renders correctly 1`] = `
.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 30px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  min-width: 30px;
  -webkit-transition: all 0.5s ease;
  transition: all 0.5s ease;
}

.c2:hover {
  background: #F6F8FB;
}

.c3 {
  border: solid #b9c1cb;
  border-width: 1px 0 0 1px;
  display: inline-block;
  padding: 4px;
  -webkit-transform: rotate(-135deg);
  -ms-transform: rotate(-135deg);
  transform: rotate(-135deg);
  -webkit-transition: all 0.5s ease;
  transition: all 0.5s ease;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  max-width: 100%;
  position: relative;
  padding: 0 10px;
}

.c1 {
  color: #7d7e7f;
  -webkit-transition: color 0.3s ease;
  transition: color 0.3s ease;
  font-size: 1rem;
  padding: 0 3px 0 10px;
}

.c1:hover {
  color: #3e4348;
}

<div
  className="c0"
  onClick={[Function]}
  style={Object {}}
>
  <span
    className="c1"
  />
  <div
    className="c2"
  >
    <i
      className="c3"
      open={false}
    />
  </div>
</div>
`;
